k256/src/arithmetic/hash2curve.rs

```diff
@@ -0,0 +1,396 @@
+            0x00, 0x00, 0x06, 0xeb,
+        ]),
+        // 0xffffffffffffffff 0xffffffffffffffff 0xffffffffffffffff 0xfffffffefffffc24
+        z: FieldElement::from_bytes_unchecked(&[
+            0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe,
+            0xff, 0xff, 0xfc, 0x24,
+        ]),
+    };
+}
+
+impl MapToCurve for ProjectivePoint {
+    type FieldElement = FieldElement;
+
+    type Output = Self;
+
+    fn map_to_curve(u: Self::FieldElement) -> Self::Output {
+        let (rx, ry) = {
+            // See section 8.7 in
+            // <https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/>
+            const C1: [u64; 4] = [
+                0xffff_ffff_bfff_ff0b,
+                0xffff_ffff_ffff_ffff,
+                0xffff_ffff_ffff_ffff,
+                0x3fff_ffff_ffff_ffff,
+            ];
+            // 0x25e9711ae8c0dadc 0x46fdbcb72aadd8f4 0x250b65073012ec80 0xbc6ecb9c12973975
+            const C2: FieldElement = FieldElement::from_bytes_unchecked(&[
+                0x25, 0xe9, 0x71, 0x1a, 0xe8, 0xc0, 0xda, 0xdc, 0x46, 0xfd, 0xbc, 0xb7, 0x2a, 0xad,
+                0xd8, 0xf4, 0x25, 0x0b, 0x65, 0x07, 0x30, 0x12, 0xec, 0x80, 0xbc, 0x6e, 0xcb, 0x9c,
+                0x12, 0x97, 0x39, 0x75,
+            ]);
+            // 0x3f8731abdd661adc 0xa08a5558f0f5d272 0xe953d363cb6f0e5d 0x405447c01a444533
+            const MAP_A: FieldElement = FieldElement::from_bytes_unchecked(&[
+                0x3f, 0x87, 0x31, 0xab, 0xdd, 0x66, 0x1a, 0xdc, 0xa0, 0x8a, 0x55, 0x58, 0xf0, 0xf5,
+                0xd2, 0x72, 0xe9, 0x53, 0xd3, 0x63, 0xcb, 0x6f, 0x0e, 0x5d, 0x40, 0x54, 0x47, 0xc0,
+                0x1a, 0x44, 0x45, 0x33,
+            ]);
+            // 0x00000000000006eb
+            const MAP_B: FieldElement = FieldElement::from_bytes_unchecked(&[
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x06, 0xeb,
+            ]);
+            // 0xffffffffffffffff 0xffffffffffffffff 0xffffffffffffffff 0xfffffffefffffc24
+            const Z: FieldElement = FieldElement::from_bytes_unchecked(&[
+                0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+                0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe,
+                0xff, 0xff, 0xfc, 0x24,
+            ]);
+
+            let tv1 = u.square(); // u^2
+            let tv3 = Z * tv1; // Z * u^2
+            let mut tv2 = tv3.square(); // tv3^2
+            let mut xd = tv2 + tv3; // tv3^2 + tv3
+            let x1n = MAP_B * (xd + FieldElement::one()); // B * (xd + 1)
+            xd = (xd * MAP_A.negate(1)).normalize(); // -A * xd
```

Comment:
I couldn't use `OsswuMap` from elliptic-curve because it's missing `normalize` in a couple of places, which it fail without. @mikelodder7 if you could make a PR for that would be amazing!

Reply:
The problem only happens when using `ct_eq` and `is_zero`. I’ve noticed that sometimes these return false without calling normalize and true when normalize is called. This sounds like a bug to me. As I’ve tested this in the `k256` crate when I add normalize calls to those two functions then everything works as expected and you can use osswu. Do you foresee a problem with this change? I can’t imagine it being one

Reply:
I'm pretty much out of my depth to answer this, @tarcieri if you could chime in here please.

Reply:
[I was worried about that before, as it were](https://github.com/RustCrypto/elliptic-curves/pull/498#issuecomment-1000376582).

You might need to use a copy-paste implementation for `k256` in order to accommodate lazy normalization. 

FWIW we do the same thing for ECDSA for similar reasons.

`p256` should be able to use the generic implementation though (of both OSSWU and ECDSA, as it were).

Reply:
That's too bad, if we don't come up with a better idea then I will "clean" this up, putting it into a separate function at least should be better.

Reply:
Unfortunately we haven’t found a good abstraction to encapsulate lazy normalization. It’s a tricky problem.

The performance gains definitely make it worth it though.

Reply:
I decided that the easiest way to clean this up was just to override the `OsswuMap` implementation.